sigstore-java/src/main/java/dev/sigstore/KeylessSigner.java

```diff
@@ -355,13 +385,43 @@ public List<Bundle> sign(List<byte[]> artifactDigests) throws KeylessSignerExcep
         throw new KeylessSignerException("Failed to validate rekor response after signing", ex);
       }
 
-      result.add(
+      var bundleBuilder =
           ImmutableBundle.builder()
               .certPath(signingCert)
               .addEntries(rekorResponse.getEntry())
               .messageSignature(
-                  MessageSignature.of(HashAlgorithm.SHA2_256, artifactDigest, signature))
-              .build());
+                  MessageSignature.of(HashAlgorithm.SHA2_256, artifactDigest, signature));
+
+      // Timestamp functionality enabled only in staging
+      if (timestampClient != null && timestampVerifier != null) {
+        var signatureDigest = Hashing.sha256().hashBytes(signature).asBytes();
+
+        var tsReq =
+            ImmutableTimestampRequest.builder()
+                .hashAlgorithm(dev.sigstore.timestamp.client.HashAlgorithm.SHA256)
+                .hash(signatureDigest)
+                .build();
+
+        TimestampResponse tsResp;
+        try {
+          tsResp = timestampClient.timestamp(tsReq);
+        } catch (TimestampException ex) {
+          throw new KeylessSignerException("Failed to generate timestamp", ex);
+        }
+
+        try {
+          timestampVerifier.verify(tsResp, signature);
+        } catch (TimestampVerificationException ex) {
+          throw new KeylessSignerException("Returned timestamp was invalid", ex);
+        }
```

Comment:
I wonder if we should check if the timestamp is truly in the validity period of the cert here as well?

Reply:
maybe we should just call KeylessVerifier.verify or something instead of doing any manual validation here?